jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java

```diff
@@ -266,10 +305,28 @@ private void validateMemScope(ResourceScope memScope) throws IllegalStateExcepti
 	/* Intended for memSegmtOfPtrToLongArgFilter that converts the memory segment
 	 * of the passed-in pointer argument to long.
 	 */
-	private final long memSegmtOfPtrToLongArg(MemorySegment argValue) throws IllegalStateException {
-		UpcallMHMetaData.validateNativeArgRetSegmentOfPtr(argValue);
+	private final long memSegmtOfPtrToLongArg(MemorySegment argValue, LinkerOptions options) throws IllegalStateException {
+		UpcallMHMetaData.validateNativeArgRetSegmentOfPtr(argValue, options);
 		addMemArgScope(argValue.scope());
-		return argValue.address();
+
+		long address = argValue.address();
+		/*[IF JAVA_SPEC_VERSION >= 22]*/
+		if (!argValue.isNative() && options.allowsHeapAccess()) {
+			HeapArgInfo info = heapArgInfo.get();
+			if (info == null) {
+				info = new HeapArgInfo(argLayoutArray.length);
+				heapArgInfo.set(info);
+			}
+			/* Store the heap argument's base object plus its offset to these arrays
+			 * which are passed to native so as to extract the heap address via macros
+			 * given address() for the heap argument is set to zero in OpenJDK.
+			 */
```

Comment:
This comment is misleading: the offset is not zero, but the offset to the first element of the array (and macros are no longer used to compute the final address given to the foreign function).

Reply:
Updated the comment as suggested above.